### PR TITLE
[control-plane-manager] default audit policy

### PR DESCRIPTION
**Pull request by 31337Ghost**
## Description
Add default audit policy. Collect logs about actions of service accounts from system namespaces (kube-system, d8-*) and about all actions taken on the resources from these namespaces.

## Why we need it and what problem does it solve?
Closes https://github.com/deckhouse/deckhouse/issues/337

## Changelog entries

```changes
module: control-plane-manager
type: feature
description: |
  Add default audit-policy.
note: |
  Due to new default audit-policy api-server component will be restarted.
```

### CODE REVIEW

File: modules/040-control-plane-manager/docs/FAQ.md [hunk 1]
```diff
@@ -136,6 +136,8 @@ Depending on the `Policy` settings and the number of requests to the **apiserver
 If **apiserver** is unable to start, you have to manually disable the `--audit-log-*` parameters in the `/etc/kubernetes/manifests/kube-apiserver.yaml` manifest and restart **apiserver** using the following command:
 ```bash
 docker stop $(docker ps | grep kube-apiserver- | awk '{print $1}')
+# or (depending on your CRI)
+crictl stopp $(crictl pods --name=kube-apiserver -q)
```
Comment (RomanenkoDenys) [inline modules/040-control-plane-manager/docs/FAQ.md:140]:
```suggestion
crictl stop $(crictl pods --name=kube-apiserver -q)
```

**Review by RomanenkoDenys — CHANGES_REQUESTED** [review]

File: modules/040-control-plane-manager/docs/FAQ_RU.md [hunk 2]
```diff
@@ -113,21 +113,21 @@ title: "Управление control plane: FAQ"
 
 **Внимание!** Операция деструктивна, полностью уничтожает консенсус и запускает etcd кластер с состояния, которое сохранилось на узле. Любые pending записи пропадут.
 
-## Как включить аудит событий?
+## Как добавить политик аудита?
 
-Если вам требуется вести учёт операций в кластере или отдебажить неожиданное поведение - для всего этого в Kubernetes предусмотрен [Auditing](https://kubernetes.io/docs/tasks/debug-application-cluster/debug-cluster/), настраиваемый через указание соответствующих [Audit Policy](https://kubernetes.io/docs/tasks/debug-application-cluster/audit/#audit-policy).
+Если вам требуется вести учёт операций в своих Namespace или отдебажить неожиданное поведение — для всего этого в Kubernetes предусмотрен [Auditing](https://kubernetes.io/docs/tasks/debug-application-cluster/debug-cluster/), настраиваемый через указание соответствующих [Audit Policy](https://kubernetes.io/docs/tasks/debug-application-cluster/audit/#audit-policy). Политики для системных Namespace включены [по–умолчанию](#аудит).
 
 На данный момент используются фиксированные параметры ротации логов:
```
Comment (apolovov) [inline modules/040-control-plane-manager/docs/FAQ_RU.md:120]: Это в конец параграфа.

File: modules/040-control-plane-manager/openapi/config-values.yaml [hunk 3]
```diff
@@ -104,7 +104,12 @@ properties:
         type: boolean
         default: false
         description: |
-          Enable [event auditing](faq.html#how-do-i-enable-event-auditing) using the configuration from the `Secret` (`audit-policy` in the `kube-system` namespace).
+          Enforce [additional audit policies](faq.html#how-do-i-add-audit-policies) using the configuration from the `Secret` (`audit-policy` in the `kube-system` namespace).
+      defaultAuditPolicyEnabled:
```
Comment (apolovov) [inline modules/040-control-plane-manager/openapi/config-values.yaml:108]: default -> basic or system.

File: modules/040-control-plane-manager/docs/README_RU.md [hunk 4]
```diff
@@ -67,3 +67,10 @@ title: Управление control plane
     - Например, `maxUsedControlPlaneVersion = 1.20`. Минимально возможная версия control plane компонентов в кластере — `1.19`.
 
 [Список поддерживаемых версий Kubernetes...](../../supported_versions.html#kubernetes)
+
+## Аудит
+В кластере с Deckhouse всегда присутствует политика для логирования событий, связанных с операциями создания, удаления и изменения:
```
Comment (apolovov) [inline modules/040-control-plane-manager/docs/README_RU.md:72]: Тут должно быть то же самое, что и в FAQ.
* что такое аудит
* результат аудита -- лог-файлик
* в декхаусе по умолчанию включен...
* кастомные правила включить — см faq

**Review by apolovov — CHANGES_REQUESTED** [review]
There is a pile of minor fixes and doc modification proposals.

File: modules/040-control-plane-manager/hooks/audit_policy_generated.go [hunk 5]
```diff
@@ -0,0 +1,93 @@
+// Code generated by "audit_policy.go" DO NOT EDIT.
```
[the remark this answers is not in the document]
Comment (apolovov) [reply modules/040-control-plane-manager/hooks/audit_policy_generated.go:1]: I would also rename the file to something like audit_policy_basic_metadata_generated.go

File: modules/040-control-plane-manager/docs/FAQ_RU.md [hunk 6]
```diff
@@ -113,21 +113,21 @@ title: "Управление control plane: FAQ"
 
 **Внимание!** Операция деструктивна, полностью уничтожает консенсус и запускает etcd кластер с состояния, которое сохранилось на узле. Любые pending записи пропадут.
 
-## Как включить аудит событий?
+## Как добавить политик аудита?
 
-Если вам требуется вести учёт операций в кластере или отдебажить неожиданное поведение - для всего этого в Kubernetes предусмотрен [Auditing](https://kubernetes.io/docs/tasks/debug-application-cluster/debug-cluster/), настраиваемый через указание соответствующих [Audit Policy](https://kubernetes.io/docs/tasks/debug-application-cluster/audit/#audit-policy).
+Если вам требуется вести учёт операций в своих Namespace или отдебажить неожиданное поведение — для всего этого в Kubernetes предусмотрен [Auditing](https://kubernetes.io/docs/tasks/debug-application-cluster/debug-cluster/), настраиваемый через указание соответствующих [Audit Policy](https://kubernetes.io/docs/tasks/debug-application-cluster/audit/#audit-policy). Политики для системных Namespace включены [по–умолчанию](#аудит).
```
[the remark this answers is not in the document]
Comment (apolovov) [reply modules/040-control-plane-manager/docs/FAQ_RU.md:118]: В FAQ оставить только "как настроить кастомную политику аудита"

File: modules/040-control-plane-manager/hooks/audit_policy.go [hunk 7]
```diff
@@ -63,28 +69,191 @@ func filterAuditSecret(unstructured *unstructured.Unstructured) (go_hook.FilterR
...
+			Resources: []audit.GroupResources{
+				{
+					Group:     "dex.coreos.com",
+					Resources: []string{"authrequests"},
+				},
+			},
+		}
+		policy.Rules = append(policy.Rules, rule)
+	}
+
```
Comment (apolovov) [inline modules/040-control-plane-manager/hooks/audit_policy.go:192]:
```suggestion
```

File: tools/audit_policy/main.go [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+	if !ok {
+		panic("cannot get caller")
+	}
+
+	dir, err := filepath.Abs(f)
+	if err != nil {
+		panic(err)
+	}
+
+	for i := 0; i < 3; i++ { // ../../
```
[the remark this answers is not in the document]
Comment (31337Ghost) [reply tools/audit_policy/main.go:59]: That code do not remove `../../` from the end of path, but returns directory path two levels upper.

File: tools/audit_policy/main.go [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+
+			*namespaces = append(*namespaces, namespace)
+		}
+
+		if filepath.Base(path) == "rbac-for-us.yaml" {
+			rbac, err := ioutil.ReadFile(path)
+			if err != nil {
+				return err
+			}
+			r := regexp.MustCompile(`apiVersion: v1
```
[the remark this answers is not in the document]
Comment (31337Ghost) [reply tools/audit_policy/main.go:115]: Yes, it's true. The strict way to retrieve all Service Account names is to compile all templates from all modules and parse them as Kubernetes resources. Regex is used here to make retrieval fast. Later, when we abandon Helm — we will have another great, easy way to refactor that.

File: modules/040-control-plane-manager/openapi/config-values.yaml (hunk 3 above)
Comment (31337Ghost) [reply modules/040-control-plane-manager/openapi/config-values.yaml:108]: ✅ `defaultAuditPolicyEnabled` became `basicAuditPolicyEnabled`

File: modules/040-control-plane-manager/hooks/audit_policy_generated.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (31337Ghost) [reply modules/040-control-plane-manager/hooks/audit_policy_generated.go:1]: I've put the full path to generator script in that comment:
```go
// Code generated by "tools/audit_policy.go" DO NOT EDIT.
```

Good point to make generated file name more suitable to it purpose. But `metadata` keyword is too broad here, so I used **targets** keyword here `audit_policy_basic_targets_generated.go`.

It should be easy to find a way to regenerate that file, by opening `tools/audit_policy.go` where you will immediately see the button: 
Screenshot 2022-01-17 at 12 49 19

File: modules/040-control-plane-manager/hooks/audit_policy.go [hunk 10]
```diff
@@ -63,28 +69,191 @@ func filterAuditSecret(unstructured *unstructured.Unstructured) (go_hook.FilterR
 }
 
 func handleAuditPolicy(input *go_hook.HookInput) error {
-	policyEnabled := input.Values.Get("controlPlaneManager.apiserver.auditPolicyEnabled")
-	if !policyEnabled.Bool() {
-		input.Values.Remove("controlPlaneManager.internal.auditPolicy")
-		return nil
+	var policy audit.Policy
+
+	if input.Values.Get("controlPlaneManager.apiserver.defaultAuditPolicyEnabled").Bool() {
```
[the remark this answers is not in the document]
Comment (31337Ghost) [reply modules/040-control-plane-manager/hooks/audit_policy.go:74]: Thank you ☺️

File: modules/040-control-plane-manager/hooks/audit_policy.go [hunk 11]
```diff
@@ -63,28 +69,191 @@ func filterAuditSecret(unstructured *unstructured.Unstructured) (go_hook.FilterR
 }
 
 func handleAuditPolicy(input *go_hook.HookInput) error {
-	policyEnabled := input.Values.Get("controlPlaneManager.apiserver.auditPolicyEnabled")
-	if !policyEnabled.Bool() {
-		input.Values.Remove("controlPlaneManager.internal.auditPolicy")
-		return nil
+	var policy audit.Policy
```
[the remark this answers is not in the document]
Comment (31337Ghost) [reply modules/040-control-plane-manager/hooks/audit_policy.go:72]: Thank you 👍

File: modules/040-control-plane-manager/docs/README.md [hunk 12]
```diff
@@ -67,3 +67,10 @@ The upgrade of a **minor version** of any control plane component is performed i
     - Suppose, `maxUsedControlPlaneVersion = 1.20`. In this case, the lowest possible version of the control plane components is `1.19`.
 
 [List of supported Kubernetes versions...](../../supported_versions.html#kubernetes)
+
+## Auditing
+In a cluster with Deckhouse, there is always a policy for logging events related to create, delete and change operations:
+- committed from the names of ServiceAccounts from the "system" Namespace `kube-system, d8- *`;
+- committed with resources in the "system" Namespace `kube-system, d8- *`.
```
[the remark this answers is not in the document]
Comment (31337Ghost) [reply modules/040-control-plane-manager/docs/README.md:74]: ✅

File: modules/040-control-plane-manager/docs/README_RU.md [hunk 13]
```diff
@@ -67,3 +67,10 @@ title: Управление control plane
...
+- совершаемых от имён сервисных аккаунтов из системных Namespace `kube-system, d8-*`;
```
[the remark this answers is not in the document]
Comment (31337Ghost) [reply modules/040-control-plane-manager/docs/README_RU.md:73]: ✅

File: modules/040-control-plane-manager/docs/README_RU.md [hunk 14]
```diff
@@ -67,3 +67,10 @@ title: Управление control plane
...
+- совершаемых с ресурсами в системных Namespace `kube-system, d8-*`.
```
[the remark this answers is not in the document]
Comment (31337Ghost) [reply modules/040-control-plane-manager/docs/README_RU.md:74]: ✅

File: modules/040-control-plane-manager/docs/README.md [hunk 15]
```diff
@@ -67,3 +67,10 @@ The upgrade of a **minor version** of any control plane component is performed i
     - Suppose, `maxUsedControlPlaneVersion = 1.20`. In this case, the lowest possible version of the control plane components is `1.19`.
 
 [List of supported Kubernetes versions...](../../supported_versions.html#kubernetes)
+
+## Auditing
+In a cluster with Deckhouse, there is always a policy for logging events related to create, delete and change operations:
+- committed from the names of ServiceAccounts from the "system" Namespace `kube-system, d8- *`;
```
[the remark this answers is not in the document]
Comment (31337Ghost) [reply modules/040-control-plane-manager/docs/README.md:73]: ✅

File: modules/040-control-plane-manager/docs/FAQ_RU.md (hunk 2 above)
Comment (31337Ghost) [reply modules/040-control-plane-manager/docs/FAQ_RU.md:120]: ✅

File: modules/040-control-plane-manager/docs/FAQ.md [hunk 16]
```diff
@@ -112,17 +112,17 @@ The control-plane-manager saves backups to `/etc/kubernetes/deckhouse/backup`. T
 
 **Caution!** This operation is unsafe and breaks the guarantees given by the consensus protocol. Note that it brings the cluster to the state that was saved on the node. Any pending entries will be lost.
 
-## How do I enable event auditing?
+## How do I add audit policies?
```
[the remark this answers is not in the document]
Comment (31337Ghost) [reply modules/040-control-plane-manager/docs/FAQ.md:115]: ✅
```markdown
## How do I configure additional audit policies?
```

File: modules/040-control-plane-manager/docs/README_RU.md (hunk 4 above)
Comment (31337Ghost) [reply modules/040-control-plane-manager/docs/README_RU.md]: ✅

File: modules/040-control-plane-manager/docs/FAQ_RU.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (31337Ghost) [reply modules/040-control-plane-manager/docs/FAQ_RU.md]: ✅

**Review by apolovov — APPROVED** [review]
👍

**Review by apolovov — APPROVED** [review]

**Pull request merged**